### ISSUE TITLE
how to send err message to client file when using passport-facebook

### ISSUE THREAD

**Issue by YoungsoonLee** [angular-fullstack/generator-angular-fullstack#1002]
Hello guys.
I have a question. I have to block duplicate email address when a user register.
So If a user use Facebook . I check the email and I want to send error message like 'already exists email address'.

So I change the code like below.

```javascript
exports.setup = function(User, config) {
    passport.use(new FacebookStrategy({
            clientID: config.facebook.clientID,
            clientSecret: config.facebook.clientSecret,
            callbackURL: config.facebook.callbackURL
        }, function(accessToken, refreshToken, profile, done) {

            User.findOne({
                    //'facebook.id': profile.id
                    'email': profile.emails[0].value
                }, function(err, user) {
                    if(err) {
                        return done(err);
                    }
                    if(!user) {
                        //check email

                        user = new User({
                            name: profile.displayName,
                            email: profile.emails[0].value,
                            role: 'user',
                            username: profile.username,
                            provider: 'facebook',
                            facebook: profile._json
                        });
                        user.save(function(err) {
                            console.log(err);
                            if(err) done(err);
                            return done(err, user);
                        });

                    } else {

                        //me change
                        //already exists email          
                        return done(null, false, {message: 'already exists email.'});

                    }

                })
        }));
};
```

but  I can't get error message in '/signuip'. I tried many times with changing code like below.
https://github.com/jaredhanson/passport-facebook/issues/93
https://github.com/jaredhanson/passport-facebook/issues/100
https://github.com/jaredhanson/passport-facebook/issues/106

but It's not working. I don't know. Could you give a advice?

**Comment by Awk34** [comment]
Do this instead:
```js
return done(null, false, {message: 'already exists email.'});
```
I don't believe there's a 3rd parameter to `done`. You should be passing that error object in the error field.

**Comment by YoungsoonLee** [comment]
Thanks . I know that I should be pass error object
but I make custom error object , I see the callback error like below

Error: Failed to lookup view "error" in views directory "D:\node project\test.ogplanet.com/server/views" at Function.app.render (D:\node project\test.ogplanet.com\node_modules\express\lib\application.js:491:17) at ServerResponse.res.render (D:\node project\test.ogplanet.com\node_modules\express\lib\response.js:777:7) at Object.handle (D:\node project\test.ogplanet.com\server\auth\facebook\index.js:37:18) at next_layer (D:\node project\test.ogplanet.com\node_modules\express\lib\router\route.js:91:15) at next_layer (D:\node project\test.ogplanet.com\node_modules\express\lib\router\route.js:87:16) at Strategy.strategy.error (D:\node project\test.ogplanet.com\node_modules\passport\lib\middleware\authenticate.js:336:9) at verified (D:\node project\test.ogplanet.com\node_modules\passport-facebook\node_modules\passport-oauth2\lib\strategy.js:177:36) at Promise.<anonymous> (D:\node project\test.ogplanet.com\server\auth\facebook\passport.js:42:18) at Promise.<anonymous> (D:\node project\test.ogplanet.com\node_modules\mongoose\node_modules\mpromise\lib\promise.js:177:8) at Promise.emit (events.js:107:17) at Promise.emit (D:\node project\test.ogplanet.com\node_modules\mongoose\node_modules\mpromise\lib\promise.js:84:38) at Promise.fulfill (D:\node project\test.ogplanet.com\node_modules\mongoose\node_modules\mpromise\lib\promise.js:97:20) at D:\node project\test.ogplanet.com\node_modules\mongoose\lib\query.js:1400:13 at model.Document.init (D:\node project\test.ogplanet.com\node_modules\mongoose\lib\document.js:254:11) at completeOne (D:\node project\test.ogplanet.com\node_modules\mongoose\lib\query.js:1398:10) at Immediate.cb (D:\node project\test.ogplanet.com\node_modules\mongoose\lib\query.js:1155:11)

Okay. thansk...I struggle to solve my problem :)

thanks a lot. 

@Awk34  you are very kind person :)
Have a nice day~!!!! and good luck :)

**Comment by Awk34** [comment]
So you solved it?

**Comment by YoungsoonLee** [comment]
not yet. I'm struggle to it :)

**Comment by Awk34** [comment]
One thing I notice is this is using incorrect slashes: `D:\node project\test.ogplanet.com/server/views`

**Comment by ethanshar** [comment]
+1
hey did someone got it solved ? i'm struggling with the same issue..

**Comment by ethanshar** [comment]
btw, there is a a 3rd parameter to done callback.
first one is the error 
second one is the user and if there's no user we should return false.
the third one is the info - which supplies information about why the authentication has failed.

basically only if u return an info it will consider as authentication failure and redirect to 'failureRedirect' u have set. 
the thing is, i want to alert the user in client side about the reason the authentication failed, but i can't seem to figure how to send this information back from the server.

**Comment by sethunath76de** [comment]
+1 struggling with exactly the same case

**Comment by simeonoff** [comment]
A little late to the party. Better late than never :) I faced a similar issue today. My app has both local registration and Facebook registration. I wanted to allow local users to link their Facebook accounts later on if they want to. I wanted to throw an error message notifying the user if an FB account was already used (linked or registered with) with another account in the database. Also, I wanted to notify the user if there was any Facebook error on callback... so here's what I did in my FB callback method:

```javascript
router.get('/facebook/callback', (req, res, next) => {
	passport.authenticate('facebook', (err) => {
		if (err) {
			console.error(err);
			return res.json({
				success: false,
				message: err.message
			});
		}
		authUser.facebookCallback(req, res);
	})(req, res, next);
});
```

This is what I did in my Facebook strategy to pass the error to the verify callback:

```javascript
return done(new Error('Facebook account already registered/linked.'), false);
```
